src/dataflow/deadlines.rs

```diff
@@ -13,105 +25,80 @@ use tokio::time::Duration;
-/// start condition on a given stream and a given timestamp (we assume a single deadline for each
-/// timestamp). Upon expiration of the deadline (defined as `duration`), the `end_condition`
-/// function is checked, and the `handler` is invoked if it was not satisfied.
-pub struct DeadlineEvent {
-    pub stream_id: StreamId,
-    pub timestamp: Timestamp,
-    pub duration: Duration,
-    pub handler: Arc<Mutex<dyn HandlerContextT>>,
-    pub end_condition: Arc<dyn CondFn>,
-}
+/*************************************************************************************************
+ * DeadlineFn: A DeadlineFn defines the duration between the start and the end condition that an *
+ * operator should be restricted to.                                                             *
+ ************************************************************************************************/
 
-impl DeadlineEvent {
-    pub fn new(
-        stream_id: StreamId,
-        timestamp: Timestamp,
-        duration: Duration,
-        handler: Arc<Mutex<dyn HandlerContextT>>,
-        end_condition: Arc<dyn CondFn>,
-    ) -> Self {
-        Self {
-            stream_id,
-            timestamp,
-            duration,
-            handler,
-            end_condition,
-        }
-    }
-}
+/// A trait that defines the deadline function. This function receives access to the State of the
+/// operator along with the current timestamp, and must calculate the time after which the deadline
+/// expires.
+pub trait DeadlineFn<S>: FnMut(&S, &Timestamp) -> Duration + Send + Sync {}
+impl<S, F: FnMut(&S, &Timestamp) -> Duration + Send + Sync> DeadlineFn<S> for F {}
 
-/// Enumerates the types of deadlines available to the user.
-pub enum Deadline {
-    TimestampDeadline(TimestampDeadline),
-}
+/*************************************************************************************************
+ * HandlerFn: A HandlerFn is invoked in case of a missed deadline.                               *
+ ************************************************************************************************/
 
-#[derive(Debug)]
-pub struct ConditionContext {
-    message_count: HashMap<(StreamId, Timestamp), usize>,
-    watermark_status: HashMap<(StreamId, Timestamp), bool>,
-}
+/// A trait that defines the handler function, which is invoked in the case of a missed deadline.
+pub trait HandlerFn<S>: FnMut(&S, &Timestamp) + Send + Sync {}
+impl<S, F: FnMut(&S, &Timestamp) + Send + Sync> HandlerFn<S> for F {}
 
-impl ConditionContext {
-    pub fn new() -> Self {
-        ConditionContext {
-            message_count: HashMap::new(),
-            watermark_status: HashMap::new(),
-        }
-    }
+/*************************************************************************************************
+ * Deadline: Define the different types of deadlines available to operators.                     *
+ ************************************************************************************************/
```

Comment:
Can we convert this to documentation (comments with 3 slashes) and move it to just above the Deadline trait?